### PR TITLE
Support multiple algorithms

### LINKED ISSUE #21: Setup support for multiple algorithms and add ALS
Currently, the only algorithm we offer is LBFGSB - it will be great to support other algorithms. And with the changes in #13, it should now be possible to handle default algorithms in an elegant and extensible way.

**Idea** change definition of `gcp` to be:
```julia
gcp(X::Array, r, loss = LeastSquaresLoss(); alg = default_algorithm(X, r, loss)) = _gcp(
    X,
    r,
    loss,
    alg,
)
```
where `default_algorithm(X, r, loss)` applies some heuristics to choose a default algorithm (e.g., ALS for `LeastSquaresLoss`, etc), and we add some algorithm types that have fields holding their parameters (with default values chosen):
+ `LBFGSB`
+ `ALS`

~~**Note:** thinking it's a good idea to remove the type restriction on `loss` to prevent duplication in `LossFunctionsExt` - losses from LossFunctions.jl will simply get routed through this signature too. Means we'll need to move that dispatch logic into `_gcp`, which is probably more sensible anyways (separates the user-friendly signature that has defaults from the dispatching logic).~~ **[UPDATE 10/12/2023: addressed in #22]**

A couple more ideas:
1. Could have `LossFunctionsExt` simply introduce a wrapper type `LossFunctionsLoss <: AbstractLoss` that wraps loss functions from `LossFunctionsExt` which could unify things even more.
2. Could also consider explicitly enumerating the actual concrete LossFunctions.jl types supported in `LossFunctionsExt`. The current code defines the domains on abstract types, which seems reasonable and applies for current losses but may not apply for all subtypes in the future. Needs some more thought.

### PR DESCRIPTION
**Pull request by dahong67**
Fixes #21.

### CODE REVIEW

**Comment by dahong67** [comment]
Needs more work on organization, will do in a separate PR.
